### PR TITLE
Update EIP-1613 to match current state of GSN 2.0 code

### PR DESCRIPTION
**Pull request by forshtat**
1. Document extraction of 'Paymaster' interface from the
'RelayRecipient' interface
2. Document using EIP-712 for signing
3. Run a spell-checker

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

**Comment by eip-automerger** [comment]
Hi! I'm a bot, and I wanted to automerge your PR, but couldn't because of the following issue(s):

 - EIP 1613 requires approval from one of (alex@tabookey.com, dror@tabookey.com, @yoav-tabookey)

File: EIPS/eip-1613.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ---
 eip:  1613
 title: Gas stations network
-author: Yoav Weiss <yoav@tabookey.com>, Dror Tirosh <dror@tabookey.com>, Alex Forshtat <alex@tabookey.com>
+author: Yoav Weiss <yoav@tabookey.com>, Dror Tirosh <dror@tabookey.com>, Alex Forshtat <forshtat1@gmail.com>
```
Comment (drortirosh) [inline EIPS/eip-1613.md]: update all emails to openeth..

File: EIPS/eip-1613.md [hunk 2]
```diff
@@ -40,52 +40,62 @@ Gas stations cannot censor transactions as long as there's at least one honest g
 
 ## Specification
 
-The system consists of a `RelayHub` singleton contract, participating contracts inheriting the `RelayRecipient` contract, a decentralized network of `Relay` nodes, a.k.a. Gas Stations, 
-and user applications (e.g. mobile or web) interacting with contracts via relays.
+The system consists of:
+* a `RelayHub` singleton contract
+* participating contracts inheriting the `RelayRecipient` contract
+* 3rd party contracts willing to provide the gas fees inheriting `Paymaster` contract (can be implemented by the `RelayRecipient` contract)
```
Comment (drortirosh) [inline EIPS/eip-1613.md]: can be _combined_ with a RelayRecipient

File: EIPS/eip-1613.md [hunk 3]
```diff
@@ -40,52 +40,62 @@ Gas stations cannot censor transactions as long as there's at least one honest g
...
+
+These are all the changes the contract needs to start accepting relayed calls. But in order for a `Relay` node to pay for a transaction, it must be compensated, so dapps are expected to also provide an implementation of the `Paymaster` contract. It can be either a separate contract or the same as `RelayRecipient`. The `Paymaster` contract must:
+
+* Maintain a small balance of ETH gas prepayment deposit in `RelayHub`. Can be paid directly by the `Paymaster` contract, or by the dapp's owner on behalf of the `Paymaster` address.
+  The `Paymaster` owner is responsible for ensuring sufficient balance for the next transactions and can stop depositing if something goes wrong, thus limiting the potential for abuse of system bugs. In DAO use-cases it will be up to the DAO logic to maintain a sufficient deposit.
+* Implement a `getGasLimits()` view function that returns a `GSNTypes.PaymasterLimits` struct. This struct contains gas limits that the Paymaster requests for its methods (discussed). The `RelayHub` will always give this exact gas limits accordingly.
```
Comment (drortirosh) [inline EIPS/eip-1613.md]: discussed => desribed below

File: EIPS/eip-1613.md [hunk 4]
```diff
@@ -40,52 +40,62 @@ Gas stations cannot censor transactions as long as there's at least one honest g
...
+  `acceptRelayedCall` is called by `RelayHub` as a view function when a `Relay` inquires it, and also during the actual transaction. Transactions are reverted if **non-zero**, and `Relay` only gets compensated for transactions (whether successful or reverted) if `acceptRelayedCall` returns **zero**. It can also optionally return `context` bytes array to communicate some data to the non-view `preRelayedCall()` and `postRelayedCall()` methods. Some examples of `acceptRelayedCall()` implementations:
     * Whitelist of trusted dapp members.
-    * Balance sheet of registered users, maintained by the dapp owner. Users pay the dapp with a credit card or other non-ETH means, and are credited in the `RelayRecipient` balance sheet. 
+    * Make a call to some method in the `RelayRecipient` and allow the Recipient to decide whether the relayed call should be accepted or not.
+    * Balance sheet of registered users, maintained by the dapp owner. Users pay the dapp with a credit card or other non-ETH means and are credited in the `RelayRecipient` balance sheet.
```
Comment (drortirosh) [inline EIPS/eip-1613.md]: in the `Paymaster` balance sheet

File: EIPS/eip-1613.md [hunk 5]
```diff
@@ -112,61 +121,61 @@ The process of sending a relayed transaction:
...
     * Value of `max_nonce` is higher than current Relay's `nonce`
-* If any of Relay's checks fail, it returns an error to sender, and doesn't proceed.
+* If any of Relay's checks fail, it returns an error to sender and doesn't proceed.
 * Relay submits the signed wrapped transaction to the blockchain.
-* Relay immediately returns the signed wrapped transaction to the sender.  This step is discussed below, in attacks/mitigations.
+* Relay immediately returns the signed wrapped transaction to the sender. This step is discussed below, in attacks/mitigations.
 * `Sender` receives the wrapped transaction and verifies that:
-    * It's a valid relay call to `RelayHub`. from Relay's address.
+    * It's a valid relay call to the `RelayHub` from Relay's address.
     * The transaction's ethereum nonce matches Relay's current nonce.
```
Comment (drortirosh) [inline EIPS/eip-1613.md]: remove the "nonce matches". it was replaced by the next line "nonce is lower"

File: EIPS/eip-1613.md [hunk 6]
```diff
@@ -260,13 +267,13 @@ so acceptRelayedCall() only returns zero for users that have enough credit, and
 With this method, the attacker can only burn its own resources, not the dapp's.
 
 A variation of this method, for free dapps (that don't charge the user, and prefer to pay for their users transactions) is to require a captcha during user creation in their web interface, 
-or to login with a Google/Facebook account, which limits the rate of the attack to the attacker's ability to open many Google/Facebook accounts. 
-Only a user that passed that process is given credit in RelayRecipient. The rate of such Sybil attack would be too low to cause any real damage.
+or to log in with a Google/Facebook account, which limits the rate of the attack to the attacker's ability to open many Google/Facebook accounts. 
+Only a user that passed that process is given credit in RelayRecipient. The rate of such a Sybil attack would be too low to cause any real damage.
```
Comment (drortirosh) [inline EIPS/eip-1613.md]: Sybil is a name of an actual woman... no "a" before a name.

File: EIPS/eip-1613.md [hunk 7]
```diff
@@ -277,24 +284,53 @@ Transactions include a nonce. RelayHub maintains a nonce (counter) for each send
...
+        bytes encodedFunction;
+        GasData gasData;
+        RelayData relayData;
+    }
+```
 
 ## Implementation
 
-A working implementation of the [**gas stations network**](https://github.com/tabookey-dev/tabookey-gasless) is being developed by **TabooKey**. It consists of `RelayHub`, `RelayRecipient`, `web3 hooks`, an implementation of a gas station inside `geth`, and sample dapps using the gas stations network.
+A working implementation of the [**gas stations network**](https://github.com/tabookey-dev/tabookey-gasless) is being developed by **TabooKey**. It consists of `RelayHub`, `RelayRecipient`, `Paymaster`, `web3 hooks`, an implementation of a gas station and sample dapps using the Gas Stations Network.
```
Comment (drortirosh) [inline EIPS/eip-1613.md]: openeth...

**Pull request closed without merging**